src/audio/mixin_mixout.c

```diff
@@ -885,103 +1086,130 @@ static int mixinout_trigger(struct comp_dev *dev, int cmd)
+	int i;
 
-	memset(params, 0, sizeof(*params));
+	memset(params, 0, sizeof(struct sof_ipc_stream_params));
 	params->channels = base_cfg->audio_fmt.channels_count;
 	params->rate = base_cfg->audio_fmt.sampling_frequency;
-	params->sample_container_bytes = base_cfg->audio_fmt.depth;
-	params->sample_valid_bytes = base_cfg->audio_fmt.valid_bit_depth;
-	params->frame_fmt = dev->ipc_config.frame_fmt;
+	params->sample_container_bytes = base_cfg->audio_fmt.depth / 8;
+	params->sample_valid_bytes = base_cfg->audio_fmt.valid_bit_depth / 8;
 	params->buffer_fmt = base_cfg->audio_fmt.interleaving_style;
-	params->buffer.size = base_cfg->ibs;
-
-	/* update each sink format based on base_cfg initialized by
-	 * host driver. There is no hw_param ipc message for ipc4, instead
-	 * all module params are built into module initialization data by
-	 * host driver based on runtime hw_params and topology setting.
-	 *
-	 * This might be not necessary for mixin as buffers between mixin and
-	 * mixout are not used (mixin writes data directly to mixout sink).
-	 * But let's keep buffer setup just in case.
-	 */
+	params->buffer.size = base_cfg->obs * 2;
+
+	audio_stream_fmt_conversion(base_cfg->audio_fmt.depth,
+				    base_cfg->audio_fmt.valid_bit_depth,
+				    &frame_fmt, &valid_fmt,
+				    base_cfg->audio_fmt.s_type);
+	params->frame_fmt = frame_fmt;
+
+	for (i = 0; i < SOF_IPC_MAX_CHANNELS; i++)
+		params->chmap[i] = (base_cfg->audio_fmt.ch_map >> i * 4) & 0xf;
+}
+
+/* params are derived from base config for ipc4 path */
+static int mixin_params(struct comp_dev *dev,
+			struct sof_ipc_stream_params *params)
+{
+	struct mixin_data *md;
+	struct list_item *blist;
+	int ret;
+
+	comp_dbg(dev, "mixin_params()");
+
+	md = comp_get_drvdata(dev);
+	base_module_cfg_to_stream_params(&md->base_cfg, params);
 
+	/* Buffers between mixins and mixouts are not used (mixin writes data directly to mixout
+	 * sink). But, anyway, let's setup these buffers properly just in case.
+	 */
 	list_for_item(blist, &dev->bsink_list) {
 		struct comp_buffer *sink;
 		struct comp_buffer __sparse_cache *sink_c;
-		int i;
+		uint16_t sink_id;
 
 		sink = buffer_from_list(blist, struct comp_buffer, PPL_DIR_DOWNSTREAM);
 		sink_c = buffer_acquire(sink);
 
-		sink_c->stream.channels = base_cfg->audio_fmt.channels_count;
-		sink_c->stream.rate = base_cfg->audio_fmt.sampling_frequency;
-		audio_stream_fmt_conversion(base_cfg->audio_fmt.depth,
-					    base_cfg->audio_fmt.valid_bit_depth,
-					    &sink_c->stream.frame_fmt,
-					    &sink_c->stream.valid_sample_fmt,
-					    base_cfg->audio_fmt.s_type);
+		sink_c->stream.channels = md->base_cfg.audio_fmt.channels_count;
 
-		sink_c->buffer_fmt = base_cfg->audio_fmt.interleaving_style;
+		/* Applying channel remapping may produce sink stream with channel count
+		 * different from source channel count.
+		 */
+		sink_id = IPC4_SRC_QUEUE_ID(sink_c->id);
+		if (sink_id < MIXIN_MAX_SINKS) {
```

Comment:
you have several cases like
```
if (a) {
	do_stuff();
} else {
	print("error");
	return -EBAD;
}
```
whereas it's easier to write
```
if (!a) {
	print("error");
	return -EBAD;
}
do_stuff();
```

Reply:
Done.